src/app/portlets/content-types/fields/content-type-fields-variables/content-type-fields-variables.component.ts

```diff
@@ -0,0 +1,135 @@
+
+    fieldVariables: FieldVariable[] = [];
+    messages: {[key: string]: string} = {};
+    form: FormGroup;
+
+    constructor(
+        private dotHttpErrorManagerService: DotHttpErrorManagerService,
+        public dotMessageService: DotMessageService,
+        private fieldVariablesService: FieldVariablesService,
+        private fb: FormBuilder
+    ) {}
+
+    ngOnInit() {
+        this.dotMessageService
+            .getMessages([
+                'contenttypes.field.variables.key_header.label',
+                'contenttypes.field.variables.value_header.label',
+                'contenttypes.field.variables.key_placeholder.label',
+                'contenttypes.field.variables.value_placeholder.label',
+                'contenttypes.field.variables.actions_header.label',
+                'contenttypes.field.variables.value_no_rows.label'
+            ])
+            .subscribe((messages: {[key: string]: string}) => {
+                this.messages = messages;
+                this.initTableData();
+            });
+        this.form = this.fb.group({
+            key: '',
+            value: ''
+        });
+    }
+
+    ngOnChanges(changes: SimpleChanges): void {
+        if (changes.field.currentValue && !changes.field.firstChange) {
+            this.initTableData();
+        }
+    }
+
+    /**
+     * Handle Create Variable event from form
+     * @memberof ContentTypeFieldsVariablesComponent
+     */
+    addVariableForm(): void {
+        const variable: FieldVariable = {
+            key: this.form.controls.key.value,
+            value: this.form.controls.value.value
+        };
+        this.saveVariable(variable);
+        this.form.setValue({ key: '', value: '' });
+    }
+
+    /**
+     * Handle Delete event
+     * @param {FieldVariable} variable
+     * @memberof ContentTypeFieldsVariablesComponent
+     */
+    deleteVariable(fieldIndex: number): void {
+        this.fieldVariablesService.deleteFieldVariables(this.field.typeId, this.field.id, this.fieldVariables[fieldIndex].id)
+            .subscribe(() => {
+                this.fieldVariables = this.fieldVariables.filter((item: FieldVariable, index: number) => index !== fieldIndex);
+            }, (err: ResponseView) => {
+                this.dotHttpErrorManagerService.handle(err).subscribe();
+            });
+    }
+
+    /**
+     * Handle Save event
+     * @param {FieldVariable} variable
+     * @memberof ContentTypeFieldsVariablesComponent
+     */
+    saveVariable(variable: FieldVariable, variableIndex?: number): void {
+        this.fieldVariablesService.saveFieldVariables(this.field.typeId, this.field.id, variable)
+            .subscribe((savedVariable: FieldVariable) => {
+                if (typeof variableIndex !== 'undefined') {
+                    this.fieldVariables = this.updateVariableCollection(savedVariable, variableIndex);
+                } else {
+                    this.fieldVariables = this.fieldVariables.concat(savedVariable);
```

Comment:
Does this trigger the change detection? do we need to trigger the change detection?

Reply:
it does trigger the change detection in the `datatable`